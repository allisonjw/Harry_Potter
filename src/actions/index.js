export const addCharacters = characters => ({
    type: 'ADD_CHARACTERS',
    characters
});

export const addHouses = houses => ({
    type: 'ADD_HOUSES',
    houses
});

export const addSpells = spells => ({
    type: 'ADD_SPELLS',
    spells
});

export const isLoading = isLoading => ({
    type: 'SET_LOADING',
    isLoading
});

export const addHouseCharacters = filteredStudents => ({
    type: 'SET_CHARACTERS',
    filteredStudents
});

export const addFavorite = (id) => ({
    type: 'ADD_FAVORITE',
    id
});
  
export const deleteFavorite = (id) => ({
    type: 'DELETE_FAVORITE',
    id
});