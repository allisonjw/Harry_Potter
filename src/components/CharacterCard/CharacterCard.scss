
.article_character-cards {
    background-color: #673ab7a3;
    border-radius: 3px;
    box-shadow: 6px 5px 7px 4px #0a021aa3;
    color: white;
    text-shadow: 2px 2px 2px black;
    text-align: center;
    grid-area: card;
    margin: 4%;
    padding: 2%;
      h3 {
          font-size: 2em;
      }
}

.favorite_icon {
  height: 30%;
    &:hover {
      cursor: pointer;
      transform: scale(1.3);
      transition: all .4s ease-in-out;
    }
}