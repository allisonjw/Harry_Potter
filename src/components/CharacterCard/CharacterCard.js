import React, { Component } from 'react';
import './CharacterCard.scss';
import { addFavorite, deleteFavorite } from '../../actions';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { PropTypes } from 'prop-types';

export class CharacterCard extends Component {
    constructor() {
        super()
        this.state = {
            favorite: !false,
        }
    };

    toggleFavorite = () => {
        const { addFavorite, deleteFavorite, favoriteCharacters, _id } = this.props;
        this.setState({ favorite : !this.state.favorite })
        favoriteCharacters.includes(_id) ? deleteFavorite(_id) : addFavorite(_id);
    }; 

    render() {
        const { favoriteCharacters, _id, name, house, dumbledoresArmy, bloodStatus, isLoading } = this.props
        const toggleMsg = favoriteCharacters.includes(_id) ? 'REMOVE FAVORITE' : 'ADD FAVORITE';
        return (
            <section className="section_character-card">
              {isLoading && <img className='loading-img' src='https://media.giphy.com/media/JonSfMu0Kzs5K9Wu7I/giphy.gif' alt='loading' />}
              {!isLoading && <article className="article_character-cards">
                    <h3>NAME:  {name}</h3>
                    <h3>HOUSE: {house}</h3>
                    <h3>DUMBLEDORE's ARMY: {dumbledoresArmy}</h3>
                    <h3>BLOOD STATUS: {bloodStatus}"</h3>
                    <h3>{toggleMsg}</h3>
                    <img src="https://media.giphy.com/media/lrJyGSrcCzLAwikcjv/giphy.gif" alt="moving goblet" className='favorite_icon' onClick={() => this.toggleFavorite()}/>
                </article>}
            </section>
        )
    };
};

export const mapStateToProps = ( state ) => ({
    favoriteCharacters: state.favoriteCharacters,
    isLoading: state.isLoading
});
  
export const mapDispatchToProps = dispatch => {
  return bindActionCreators({ 
      addFavorite: id => dispatch( addFavorite(id) ),
      deleteFavorite: id => dispatch( deleteFavorite(id) ),
  }, dispatch)
};

export default connect(mapStateToProps, mapDispatchToProps)(CharacterCard);

CharacterCard.propTypes = {
    isLoading: PropTypes.string,
    filteredCharacters: PropTypes.array,
    favoriteCharacters: PropTypes.array,
    addFavorite: PropTypes.string,
    deleteFavorite: PropTypes.string
  
};