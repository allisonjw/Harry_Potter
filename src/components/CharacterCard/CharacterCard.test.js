import React from 'react';
import { shallow } from 'enzyme';
import { CharacterCard, mapStateToProps, mapDispatchToProps } from './CharacterCard';
import { addFavorite, deleteFavorite } from '../../actions';

describe('CharacterCard', () => {
 
    describe('Characters', () => {
        let mockFavoriteCharacters = [
              { 
                _id: "5a0fa4daae5bc100213c232e",
                name: "Hannah Abbott",
                role: "student",
                house: "Hufflepuff",
                school: "Hogwarts School of Witchcraft and Wizardry",
                __v: 0,
                ministryOfMagic: false,
                orderOfThePhoenix: false,
                dumbledoresArmy: true,
                deathEater: false,
                bloodStatus: "half-blood",
                species: "human"
             }
           ];
    
           let mockCharacterCard = [
              { 
                _id: "5a0fa4daae5bc100213c232e",
                name: "Hannah Abbott",
                role: "student",
                house: "Hufflepuff",
                school: "Hogwarts School of Witchcraft and Wizardry",
                __v: 0,
                ministryOfMagic: false,
                orderOfThePhoenix: false,
                dumbledoresArmy: true,
                deathEater: false,
                bloodStatus: "half-blood",
                species: "human"
              },
              { 
                _id: "5a0fa54aae5bc100213c232f",
                name: "Bathsheda Babbling",
                role: "Professor, Ancient Runes",
                school: "Hogwarts School of Witchcraft and Wizardry",
                __v: 0,
                ministryOfMagic: false,
                orderOfThePhoenix: false,
                dumbledoresArmy: true,
                deathEater: false,
                bloodStatus: "unknown",
                species: "human"
              } 
           ];

        let wrapper = shallow( <CharacterCard 
            isLoading='false'
            CharacterCard={mockCharacterCard}
            favoriteCharacters={mockFavoriteCharacters}
            addFavorite='5b74f3653228320021ab628b'
            deleteFavorite='5b74f3653228320021ab628b'
         />);

         it('should match snapshot with all information passing in correctly', () => {
             expect(wrapper).toMatchSnapshot();
         });

         it.skip('should call toggleFavorite when image is clicked', () => {
             wrapper.instance().toggleFavorite = jest.fn();
             wrapper.find('.favorite_icon').simulate('click');
             expect(wrapper.instance().toggleFavorite).toHaveBeenCalled();
         });
    });

    describe('mapStateToProps', () => {
        it('should return all state with their properties', () => {
            const mockState = {
                "isLoading": 'false',
                "favoriteCharacters": [   
                  { 
                    _id: "5a0fa4daae5bc100213c232e",
                    name: "Hannah Abbott",
                    role: "student",
                    house: "Hufflepuff",
                    school: "Hogwarts School of Witchcraft and Wizardry",
                    __v: 0,
                    ministryOfMagic: false,
                    orderOfThePhoenix: false,
                    dumbledoresArmy: true,
                    deathEater: false,
                    bloodStatus: "half-blood",
                    species: "human"
                   }
                 ]
             };

             const expected = {
                "isLoading": 'false',
                "favoriteCharacters": [
                   { 
                    _id: "5a0fa4daae5bc100213c232e",
                    name: "Hannah Abbott",
                    role: "student",
                    house: "Hufflepuff",
                    school: "Hogwarts School of Witchcraft and Wizardry",
                    __v: 0,
                    ministryOfMagic: false,
                    orderOfThePhoenix: false,
                    dumbledoresArmy: true,
                    deathEater: false,
                    bloodStatus: "half-blood",
                    species: "human"
                  }
                ]
             };
             const mappedProps = mapStateToProps(mockState);
             expect(mappedProps).toEqual(expected);
        });
    });

    describe('mapDispatchToProps', () => {
        it('calls dispatch with an addFavorite action when toggleFavorite is called', () => {
            const mockDispatch = jest.fn();
            const actionToDispatch = addFavorite('5b74f3653228320021ab628b');

            const mappedProps = mapDispatchToProps(mockDispatch);
            mappedProps.addFavorite('5b74f3653228320021ab628b');

            expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
          });

          it('calls dispatch with an deleteFavorite action when toggleFavorite is called', () => {
            const mockDispatch = jest.fn();
            const actionToDispatch = deleteFavorite('5b74f3653228320021ab628b');
            const mappedProps = mapDispatchToProps(mockDispatch);

            mappedProps.deleteFavorite('5b74f3653228320021ab628b');
            expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
          });
    });
});