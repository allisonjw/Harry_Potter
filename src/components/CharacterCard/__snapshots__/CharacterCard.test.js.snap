// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterCard Characters should match snapshot with all information passing in correctly 1`] = `
<section
  className="section_character-card"
>
  <img
    alt="loading"
    className="loading-img"
    src="https://media.giphy.com/media/JonSfMu0Kzs5K9Wu7I/giphy.gif"
  />
</section>
`;
